File: tyhp-lang/vscode/src/workspace/selectOwner.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { selectOwner } from "./selectOwner";

test("single candidate is the owner", () => {
    const owner = selectOwner("/repo/runtime/packages/core/tyhp_src/Type.tyhp", [
        { projectFilePath: "/repo/runtime/packages/core/tyhp.json", projectDir: "/repo/runtime/packages/core" },
    ]);
    assert.equal(owner?.projectFilePath, "/repo/runtime/packages/core/tyhp.json");
});

test("empty candidates yields no owner", () => {
    assert.equal(selectOwner("/repo/a.tyhp", []), undefined);
});

test("nearest ancestor wins over a non-ancestor that also matches", () => {
    const owner = selectOwner("/repo/app/pkg/src/Main.tyhp", [
        { projectFilePath: "/repo/runtime/packages/core/tyhp.json", projectDir: "/repo/runtime/packages/core" },
        { projectFilePath: "/repo/app/pkg/tyhp.json", projectDir: "/repo/app/pkg" },
        { projectFilePath: "/repo/app/tyhp.json", projectDir: "/repo/app" },
    ]);
    assert.equal(owner?.projectFilePath, "/repo/app/pkg/tyhp.json");
});

test("among ancestors, the nearest (deepest) folder wins", () => {
    const owner = selectOwner("/ws/lib/mod/src/a.tyhp", [
        { projectFilePath: "/ws/tyhp.json", projectDir: "/ws" },
        { projectFilePath: "/ws/lib/tyhp.json", projectDir: "/ws/lib" },
        { projectFilePath: "/ws/lib/mod/tyhp.json", projectDir: "/ws/lib/mod" },
    ]);
    assert.equal(owner?.projectFilePath, "/ws/lib/mod/tyhp.json");
});

test("when no ancestor matches, nearest by path hops wins", () => {
    const file = "/repo/runtime/php-extensions/php8.2.9/standard/strings.tyhpdef";
    const owner = selectOwner(file, [
        { projectFilePath: "/repo/runtime/packages/core/tyhp.json", projectDir: "/repo/runtime/packages/core" },
        { projectFilePath: "/repo/other/far/away/tyhp.json", projectDir: "/repo/other/far/away" },
    ]);
    assert.equal(owner?.projectFilePath, "/repo/runtime/packages/core/tyhp.json");
});

test("equal hops: shortest tyhp.json path, then lexicographic", () => {
    const file = "/repo/runtime/php-extensions/php8.2.9/standard/strings.tyhpdef";
    const owner = selectOwner(file, [
        { projectFilePath: "/repo/runtime/packages/core/tyhp.json", projectDir: "/repo/runtime/packages/core" },
        { projectFilePath: "/repo/runtime/packages/async/tyhp.json", projectDir: "/repo/runtime/packages/async" },
        { projectFilePath: "/repo/runtime/packages/lambda/tyhp.json", projectDir: "/repo/runtime/packages/lambda" },
    ]);
    assert.equal(owner?.projectFilePath, "/repo/runtime/packages/core/tyhp.json");
});

test("lexicographic tie-break when path lengths are equal", () => {
    const file = "/shared/x.tyhpdef";
    const owner = selectOwner(file, [
        { projectFilePath: "/b/tyhp.json", projectDir: "/b" },
        { projectFilePath: "/a/tyhp.json", projectDir: "/a" },
    ]);
    assert.equal(owner?.projectFilePath, "/a/tyhp.json");
});

File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/workspace/SelectOwnerTest.kt
package com.tyhp.lang.workspace

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class SelectOwnerTest {
    @Test
    fun `single candidate is the owner`() {
        val owner = selectOwner(
            "/repo/runtime/packages/core/tyhp_src/Type.tyhp",
            listOf(SimpleOwner("/repo/runtime/packages/core/tyhp.json", "/repo/runtime/packages/core")),
        )
        assertEquals("/repo/runtime/packages/core/tyhp.json", owner?.projectFilePath)
    }

    @Test
    fun `empty candidates yields no owner`() {
        assertNull(selectOwner("/repo/a.tyhp", emptyList()))
    }

    @Test
    fun `nearest ancestor wins`() {
        val owner = selectOwner(
            "/repo/app/pkg/src/Main.tyhp",
            listOf(
                SimpleOwner("/repo/runtime/packages/core/tyhp.json", "/repo/runtime/packages/core"),
                SimpleOwner("/repo/app/pkg/tyhp.json", "/repo/app/pkg"),
                SimpleOwner("/repo/app/tyhp.json", "/repo/app"),
            ),
        )
        assertEquals("/repo/app/pkg/tyhp.json", owner?.projectFilePath)
    }

    @Test
    fun `when no ancestor matches, nearest by path hops wins`() {
        val owner = selectOwner(
            "/repo/runtime/php-extensions/php8.2.9/standard/strings.tyhpdef",
            listOf(
                SimpleOwner("/repo/runtime/packages/core/tyhp.json", "/repo/runtime/packages/core"),
                SimpleOwner("/repo/other/far/away/tyhp.json", "/repo/other/far/away"),
            ),
        )
        assertEquals("/repo/runtime/packages/core/tyhp.json", owner?.projectFilePath)
    }

    @Test
    fun `equal hops shortest path then lexicographic`() {
        val owner = selectOwner(
            "/repo/runtime/php-extensions/php8.2.9/standard/strings.tyhpdef",
            listOf(
                SimpleOwner("/repo/runtime/packages/core/tyhp.json", "/repo/runtime/packages/core"),
                SimpleOwner("/repo/runtime/packages/async/tyhp.json", "/repo/runtime/packages/async"),
                SimpleOwner("/repo/runtime/packages/lambda/tyhp.json", "/repo/runtime/packages/lambda"),
            ),
        )
        assertEquals("/repo/runtime/packages/core/tyhp.json", owner?.projectFilePath)
    }

    @Test
    fun `lexicographic tie-break when path lengths are equal`() {
        val owner = selectOwner(
            "/shared/x.tyhpdef",
            listOf(
                SimpleOwner("/b/tyhp.json", "/b"),
                SimpleOwner("/a/tyhp.json", "/a"),
            ),
        )
        assertEquals("/a/tyhp.json", owner?.projectFilePath)
    }
}
